src/lib/jsonParseV2/parseRegistration.cpp

```diff
@@ -0,0 +1,367 @@
+  {
+    *errorStringP = "/dataProvided/ is empty";
+    return false;
+  }
+
+  if (dataProvided.HasMember("entities"))
+  {
+    bool b = parseEntitiesVector(ciP, &dataProvidedP->entities, dataProvided["entities"], errorStringP);
+    if (b == false)
+    {
+      return false;
+    }
+  }
+
+  if (dataProvided.HasMember("attrs"))
+  {
+    if (!parseStringVector(&dataProvidedP->attributes, dataProvided["attrs"], "attrs", true, errorStringP))
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* httpParse -
+*/
+static bool httpParse(ngsiv2::Http* httpP, const rapidjson::Value& jsonHttp, const std::string& fieldName, std::string* errorStringP)
+{
+  if (!jsonHttp.IsObject())
+  {
+    *errorStringP = "/" + fieldName + "/ must be a JSON object";
+    return false;
+  }
+
+  if (!jsonHttp.HasMember("url"))
+  {
+    *errorStringP = "/" + fieldName + "/ must contain a field /url/";
+    return false;
+  }
+
+  const rapidjson::Value& url = jsonHttp["url"];
+
+  if (!url.IsString())
+  {
+    *errorStringP = "/url/ field of /" + fieldName + "/ must be a JSON string";
+    return false;
+  }
+
+  httpP->url = url.GetString();
+
+  if (httpP->url == "")
+  {
+    *errorStringP = "/url/ field of /" + fieldName + "/ cannot be the empty string";
+    return false;
+  }
+
+  std::string  host;
+  int          port;
+  std::string  path;
+  std::string  protocol;
+
+  if (parseUrl(httpP->url, host, port, path, protocol) == false)
+  {
+    *errorStringP = "/url/ field of /" + fieldName + "/ is invalid";
+    return false;
+  }
+
+  return true;
+}
+
+
+
+/* ****************************************************************************
+*
+* providerParse -
+*/
+static bool providerParse(ConnectionInfo* ciP, ngsiv2::Provider* providerP, const rapidjson::Value& provider, std::string* errorStringP)
+{
+  bool b;
+
+  if (!provider.IsObject())
+  {
+    *errorStringP = "/provider/ must be a JSON object";
+    return false;
+  }
+
+  if (provider.ObjectEmpty())
+  {
+    *errorStringP = "/provider/ empty";
+    return false;
+  }
+
+  if (provider.HasMember("http"))
+  {
+    b = httpParse(&providerP->http, provider["http"], "http", errorStringP);
+    if (b == false)
+    {
+      return false;
+    }
+  }
+  else
+  {
+    *errorStringP = "/provider/ must contain a field /http/";
+    return false;
+  }
+
+  if (provider.HasMember("supportedForwardingMode"))
+  {
+    const rapidjson::Value& supportedForwardingMode = provider["supportedForwardingMode"];
+    if (!supportedForwardingMode.IsString())
+    {
+      *errorStringP = "/supportedForwardingMode/ must be a JSON string";
+      return false;
+    }
+
+    providerP->supportedForwardingMode = supportedForwardingMode.GetString();
+    
+    //
+    // Valid strings: ngsiV2, legacy, all
+    //
+    if ((providerP->supportedForwardingMode != "ngsiV2") &&  // FIXME PR: what are the exact values?
```

Comment:
Not sure of understanding what do you mean by this FIXME PR comment, sorry...

Reply:
It's about the names of the 'forwarding modes'.
  Are the valid values: "ngsiV2", "legacy2, "all" ?

Reply:
Let's check http://telefonicaid.github.io/fiware-orion/api/v2/latest...

```
* `supportedForwardingMode`: It is used to convey the forwarding mode supported by this context provider. By default `all`. Allowed values are:
 * `none`: This provider does not support request forwarding.
 *  `query`: This provider only supports request forwarding to query data.
 *  `update`: This provider only supports request forwarding to update data.
 *  `all`: This provider supports both query and update forwarding requests. (Default value)
```

Mabye an enum will be a good idea for that (which the appropiate FIXME regarding yet unused values, i.e. all except the one corresponding to "all")

Reply:
Sure, an enum is good. I'll use these very texts (in the parsing state, the supportedForwardingMode always is a string) and translate to enum constants

Reply:
Fixed in 401b10d8a